File: solver/board.py
# board.py
from typing import List

rows = 8
cols = 6
max_steps = 40


class Position:
    def __init__(self, row=0, col=0):
        self.row = row
        self.col = col

    def __ne__(self, other):
        return self.row != other.row or self.col != other.col


POSITION_NONE = Position(15, 15)


class Field:
    def __init__(self, color="g", modifier="0"):
        self.color = color
        self.modifier = modifier
        self.onlyReachableFrom = POSITION_NONE

    def is_color(self, c):
        return c in "rgbod"

    def is_clickable(self):
        return (
            self.is_static_arrow() or self.is_rotating_arrow() or self.modifier in "FB"
        )

    def is_static_arrow(self):
        return self.modifier in "LRUD"

    def is_rotating_arrow(self):
        return self.modifier in "wsax"

    def is_correct(self):
        if not self.is_color(self.color):
            return True
        if self.is_color(self.modifier):
            return self.color == self.modifier
        return self.modifier != "0"


class MoveSequence:
    def __init__(self):
        self.moves: List[Position] = []
        self.n = 0

    def add(self, row, col):
        self.moves.append(Position(row, col))
        self.n += 1

    def __str__(self):
        return ",".join(f"{chr(m.col + ord('A'))}{m.row + 1}" for m in self.moves)


class Board:
    def __init__(self):
        self.fields = [[Field() for _ in range(cols)] for _ in range(rows)]
        self.move_sequence = MoveSequence()
        self.has_bombs = False

    def click(self, row, col):
        self.move_sequence.add(row, col)
        f = self.fields[row][col]

        # Arrow moves
        def fill(dr, dc, r, c, color):
            r += dr
            c += dc
            if not (0 <= r < rows and 0 <= c < cols):
                return False
            from_color = None
            to_color = None
            if self.fields[r][c].modifier == color:
                from_color, to_color = color, "0"
            elif self.fields[r][c].modifier == "0":
                from_color, to_color = "0", color
            else:
                return False
            changed = False
            while (
                0 <= r < rows
                and 0 <= c < cols
                and self.fields[r][c].modifier == from_color
            ):
                self.fields[r][c].modifier = to_color
                r += dr
                c += dc
                changed = True
            return changed

        # Flood fill
        def flood(r, c, from_color, to_color):
            if not (0 <= r < rows and 0 <= c < cols):
                return False
            if self.fields[r][c].modifier != from_color:
                return False
            self.fields[r][c].modifier = to_color
            changed = True
            changed |= flood(r + 1, c, from_color, to_color)
            changed |= flood(r - 1, c, from_color, to_color)
            changed |= flood(r, c + 1, from_color, to_color)
            changed |= flood(r, c - 1, from_color, to_color)
            return changed

        mod = f.modifier
        if mod == "U":
            return fill(-1, 0, row, col, f.color)
        elif mod == "D":
            return fill(1, 0, row, col, f.color)
        elif mod == "L":
            return fill(0, -1, row, col, f.color)
        elif mod == "R":
            return fill(0, 1, row, col, f.color)
        elif mod == "F":
            changed = False
            changed |= flood(row + 1, col, "0", f.color)
            changed |= flood(row - 1, col, "0", f.color)
            changed |= flood(row, col + 1, "0", f.color)
            changed |= flood(row, col - 1, "0", f.color)
            if not changed:
                changed |= flood(row + 1, col, f.color, "0")
                changed |= flood(row - 1, col, f.color, "0")
                changed |= flood(row, col + 1, f.color, "0")
                changed |= flood(row, col - 1, f.color, "0")
            return changed
        elif mod == "B":
            for dr in range(3):
                for dc in range(3):
                    r, c = row - 1 + dr, col - 1 + dc
                    if (
                        0 <= r < rows
                        and 0 <= c < cols
                        and self.fields[r][c].modifier != "X"
                    ):
                        self.fields[r][c].modifier = f.color
            return True
        elif mod == "w":
            fill(-1, 0, row, col, f.color)
            f.modifier = "x"
            return True
        elif mod == "s":
            fill(1, 0, row, col, f.color)
            f.modifier = "a"
            return True
        elif mod == "a":
            fill(0, -1, row, col, f.color)
            f.modifier = "w"
            return True
        elif mod == "x":
            fill(0, 1, row, col, f.color)
            f.modifier = "s"
            return True
        else:
            print("Unknown modifier", mod)
        return False

    def is_solved(self):
        return all(f.is_correct() for row in self.fields for f in row)

    @classmethod
    def from_strings(cls, color_str, modifier_str):
        color_str = color_str.replace(" ", "")
        modifier_str = modifier_str.replace(" ", "")
        b = cls()
        small_board = len(color_str) == 5 * 6
        for r in range(rows):
            for c in range(cols):
                f = b.fields[r][c]
                if small_board and (r >= 6 or c >= 5):
                    f.color = "0"
                    f.modifier = "X"
                    continue
                idx = r * (5 if small_board else 6) + c
                if idx >= len(color_str) or idx >= len(modifier_str):
                    f.color = "0"
                    f.modifier = "X"
                else:
                    f.color = color_str[idx]
                    f.modifier = modifier_str[idx]
        return b

    def display(self):
        for row in range(rows):
            for col in range(cols):
                f = self.fields[row][col]

                if f.color == "0":
                    # Transparent: just print spaces
                    print("  ", end="")
                else:
                    color_code = {
                        "r": 41,  # Red
                        "g": 42,  # Green
                        "b": 44,  # Blue
                        "o": 43,  # Yellow/Orange
                        "d": 45,  # Magenta/Purple
                    }.get(f.color, 40)
                    print(f"\033[{color_code}m  \033[0m", end="")
            print()

    def hash(self):
        """
        Compute a hash of the board’s current state using colors and modifiers.
        """
        flat = []
        for row in self.fields:
            for field in row:
                flat.append((field.color, field.modifier))
        return hash(tuple(flat))

    def copy(self):
        new_board = Board()
        for r in range(rows):
            for c in range(cols):
                f_src = self.fields[r][c]
                f_dst = new_board.fields[r][c]
                f_dst.color = f_src.color
                f_dst.modifier = f_src.modifier
                f_dst.onlyReachableFrom = f_src.onlyReachableFrom
        new_board.move_sequence = MoveSequence()
        for m in self.move_sequence.moves:
            new_board.move_sequence.add(m.row, m.col)
        new_board.has_bombs = self.has_bombs
        return new_board

File: solver/board.js
class Position {
  constructor(row = 0, col = 0) {
    this.row = row;
    this.col = col;
  }

  notEquals(other) {
    return this.row !== other.row || this.col !== other.col;
  }
}

const POSITION_NONE = new Position(15, 15);

class Field {
  constructor(color = "g", modifier = "0") {
    this.color = color;
    this.modifier = modifier;
    this.onlyReachableFrom = POSITION_NONE;
  }

  isColor(c) {
    return "rgbod".includes(c);
  }

  isClickable() {
    return this.isStaticArrow() || this.isRotatingArrow() || "FB".includes(this.modifier);
  }

  isStaticArrow() {
    return "LRUD".includes(this.modifier);
  }

  isRotatingArrow() {
    return "wsax".includes(this.modifier);
  }

  isCorrect() {
    if (!this.isColor(this.color)) return true;
    if (this.isColor(this.modifier)) return this.color === this.modifier;
    return this.modifier !== "0";
  }
}

class MoveSequence {
  constructor() {
    this.moves = [];
    this.n = 0;
  }

  add(row, col) {
    this.moves.push(new Position(row, col));
    this.n++;
  }

  toString() {
    return this.moves.map(m => String.fromCharCode(m.col + 65) + (m.row + 1)).join(",");
  }
}

class Board {
  constructor(rows = 8, cols = 6) {
    this.rows = rows;
    this.cols = cols;
    this.fields = Array.from({ length: rows }, () =>
      Array.from({ length: cols }, () => new Field())
    );
    this.moveSequence = new MoveSequence();
    this.hasBombs = false;
  }

  click(row, col) {
    this.moveSequence.add(row, col);
    const f = this.fields[row][col];

    const fill = (dr, dc, r, c, color) => {
      r += dr; c += dc;
      if (!(0 <= r && r < this.rows && 0 <= c && c < this.cols)) return false;
      let fromColor, toColor;
      const mod = this.fields[r][c].modifier;
      if (mod === color) { fromColor = color; toColor = "0"; }
      else if (mod === "0") { fromColor = "0"; toColor = color; }
      else return false;
      let changed = false;
      while (0 <= r && r < this.rows && 0 <= c && c < this.cols && this.fields[r][c].modifier === fromColor) {
        this.fields[r][c].modifier = toColor;
        r += dr; c += dc;
        changed = true;
      }
      return changed;
    };

    const flood = (r, c, fromColor, toColor) => {
      if (!(0 <= r && r < this.rows && 0 <= c && c < this.cols)) return false;
      if (this.fields[r][c].modifier !== fromColor) return false;
      this.fields[r][c].modifier = toColor;
      let changed = true;
      changed |= flood(r + 1, c, fromColor, toColor);
      changed |= flood(r - 1, c, fromColor, toColor);
      changed |= flood(r, c + 1, fromColor, toColor);
      changed |= flood(r, c - 1, fromColor, toColor);
      return changed;
    };

    const mod = f.modifier;
    switch (mod) {
      case "U": return fill(-1, 0, row, col, f.color);
      case "D": return fill(1, 0, row, col, f.color);
      case "L": return fill(0, -1, row, col, f.color);
      case "R": return fill(0, 1, row, col, f.color);
      case "F":
        let changed = false;
        changed |= flood(row + 1, col, "0", f.color);
        changed |= flood(row - 1, col, "0", f.color);
        changed |= flood(row, col + 1, "0", f.color);
        changed |= flood(row, col - 1, "0", f.color);
        if (!changed) {
          changed |= flood(row + 1, col, f.color, "0");
          changed |= flood(row - 1, col, f.color, "0");
          changed |= flood(row, col + 1, f.color, "0");
          changed |= flood(row, col - 1, f.color, "0");
        }
        return changed;
      case "B":
        for (let dr = 0; dr < 3; dr++) {
          for (let dc = 0; dc < 3; dc++) {
            const r = row - 1 + dr;
            const c = col - 1 + dc;
            if (0 <= r && r < this.rows && 0 <= c && c < this.cols && this.fields[r][c].modifier !== "X") {
              this.fields[r][c].modifier = f.color;
            }
          }
        }
        return true;
      case "w": fill(-1, 0, row, col, f.color); f.modifier = "x"; return true;
      case "s": fill(1, 0, row, col, f.color); f.modifier = "a"; return true;
      case "a": fill(0, -1, row, col, f.color); f.modifier = "w"; return true;
      case "x": fill(0, 1, row, col, f.color); f.modifier = "s"; return true;
      default: console.log("Unknown modifier", mod); return false;
    }
  }

  isSolved() {
    return this.fields.flat().every(f => f.isCorrect());
  }

  static fromStrings(colorStr, modifierStr) {
    // Determine rows/cols from XML
    const colorLines = colorStr.trim().split(/\s+/);
    const modifierLines = modifierStr.trim().split(/\s+/);

    const rows = Math.max(colorLines.length, modifierLines.length);
    const cols = Math.max(...colorLines.map(l => l.length), ...modifierLines.map(l => l.length));

    // Flatten
    const colorClean = colorLines.join("");
    const modifierClean = modifierLines.join("");

    const b = new Board(rows, cols);
    for (let r = 0; r < rows; r++) {
      for (let c = 0; c < cols; c++) {
        const f = b.fields[r][c];
        const idx = r * cols + c;
        if (idx >= colorClean.length || idx >= modifierClean.length) {
          f.color = "0";
          f.modifier = "X";
        } else {
          f.color = colorClean[idx];
          f.modifier = modifierClean[idx];
        }
      }
    }
    return b;
  }

  display() {
    for (let r = 0; r < this.rows; r++) {
      let line = "";
      for (let c = 0; c < this.cols; c++) {
        const f = this.fields[r][c];
        if (f.color === "0") line += "  ";
        else {
          const code = { r: 41, g: 42, b: 44, o: 43, d: 45 }[f.color] || 40;
          line += `\x1b[${code}m  \x1b[0m`;
        }
      }
      console.log(line);
    }
  }

  hash() {
    return JSON.stringify(this.fields.map(r => r.map(f => [f.color, f.modifier])));
  }

  copy() {
    const b = new Board(this.rows, this.cols);
    for (let r = 0; r < this.rows; r++) {
      for (let c = 0; c < this.cols; c++) {
        const fSrc = this.fields[r][c];
        const fDst = b.fields[r][c];
        fDst.color = fSrc.color;
        fDst.modifier = fSrc.modifier;
        fDst.onlyReachableFrom = fSrc.onlyReachableFrom;
      }
    }
    b.moveSequence.moves = this.moveSequence.moves.map(m => new Position(m.row, m.col));
    b.moveSequence.n = this.moveSequence.n;
    b.hasBombs = this.hasBombs;
    return b;
  }
}

module.exports = { Board, Field, MoveSequence, Position, POSITION_NONE };
